lexicon/providers/auto.py

```diff
@@ -0,0 +1,173 @@
+# Indeed we want to delegate every parameter/method call to the delegate provider
+# but __getattr__ is called only if the parameter/method cannot be found in the
+# current Provider hierarchy. If it is object, it will be the case for every relevant
+# call in the Lexicon library.
+class Provider(object):
+    """
+    Implementation of the provider 'auto'.
+    For the given domain, it will resolve the actual Provider class to use by inspecting the 
+    nameservers declared for the domain, using declared nameservers domain in each Lexicon DNS
+    provider.
+    Any call upon instantiation will be delegated to the resolved Provider.
+    Any command line parameter or environment variable passed to Lexicon will be transfered to
+    the resolved provider if it respect the naming convention: --[provider]-[parameter_name] for a
+    command line parameter, or LEXICON_[PROVIDER]_PARAMETER_NAME for a environment variable.
+    """
+    def __init__(self, options, engine_overrides=None):
+        self.domain = options.get('domain')
+        self.delegate = None
+        self.options = options
+        self.engine_overrides = engine_overrides
+
+    def authenticate(self):
+        """
+        Launch the authentication process: for 'auto' provider, it means first to find the relevant
+        provider, then call its authenticate() method. Almost every subsequent operation will then 
+        be delegated to that provider.
+        """
+        mapping_override = self.options.get('mapping_override')
+        mapping_override_processed = {}
+        if mapping_override:
+            for one_mapping in mapping_override.split(','):
+                one_mapping_processed = one_mapping.split(':')
+                mapping_override_processed[one_mapping_processed[0]] = one_mapping_processed[1]
+
+        override_provider = mapping_override_processed.get(self.domain)
+        if override_provider:
+            provider = [element for element in AVAILABLE_PROVIDERS if element.__name__ == override_provider][0]
+            LOGGER.info('Provider authoritatively mapped for domain %s: %s.', self.domain, provider.__name__)
+        else:
+            (provider_name, provider_module) = _relevant_provider_for_domain(self.domain)
+            LOGGER.info('Provider discovered for domain %s: %s.', self.domain, provider_name)
+
+        new_options = env_auth_options(provider_name)
+        for key, value in self.options.items():
+            target_prefix = 'auto_{0}_'.format(provider_name)
+            if key.startswith(target_prefix):
+                new_options[re.sub('^{0}'.format(target_prefix), '', key)] = value
+            if not key.startswith('auto_'):
+                new_options[key] = value
+
+        new_options['provider_name'] = provider_name
+
+        self.delegate = provider_module.Provider(new_options, self.engine_overrides)
+        self.delegate.authenticate()
+
+    def __getattr__(self, attr_name):
```

Comment:
this is smart, good work 👍

Reply:
Thanks ! For the record, you could ask why it did not placed the provider proxy resolution logic in the constructor. It is because of Certbot. Last time I pushed some advanced logic in the constructor, Certbot tests could fail because their mock are enabled only on the provider methods, not its constructor.


Quite related in the Lexicon tests. I put some effort to not rely on an actual execution of the `nslookup` binary by mocking the relevant method call, and make the tests more portable. The only actual execution of `nslookup` is done on an explicit test designed for that, and here again I skip it automatically for known situations where it cannot work correctly (no network on the test system for instance).